qemu/tests/virtio_serial_hotplug_max_chardevs.py

```diff
@@ -0,0 +1,108 @@
+
+        for serial_port in serials:
+            port_params = params.object_params(serial_port)
+            if not port_params['serial_type'].startswith('virtserial'):
+                continue
+            logging.info("transfer data with port %s" % serial_port)
+            params['file_transfer_serial_port'] = serial_port
+            transfer_data(params, vm, sender='both')
+
+    def run_bg_test():
+        """
+        Set the operation of transferring data as background
+        :return: return the background case thread if it's successful;
+                 else raise error
+        """
+        stress_thread = utils_misc.InterruptedThread(run_serial_data_transfer)
+        stress_thread.start()
+        if not utils_misc.wait_for(lambda: driver_in_use.check_bg_running(
+                vm, params), check_bg_timeout, 0, 1):
+            test.fail("Backgroud test is not alive!")
+        return stress_thread
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    os_type = params["os_type"]
+    check_bg_timeout = float(params.get('check_bg_timeout', 120))
+    num_chardev = int(params.get("numberic_chardev"))
+    num_serial_ports = int(params.get("virtio_serial_ports"))
+    for i in range(1, num_chardev):
+        params["extra_chardevs"] += ' channel%d' % i
+        serial_name = 'port%d' % i
+        params['extra_serials'] = '%s %s' % (params.get('extra_serials', ''),
+                                             serial_name)
+        params['serial_type_%s' % serial_name] = "virtserialport"
+    char_devices = add_chardev(vm, params)
+    serials = params.objects('extra_serials')
+    buses, serial_devices = get_buses_and_serial_devices(
+        vm, params, char_devices, serials)
+    vm.devices.simple_hotplug(buses[0], vm.monitor)
+    for i in range(0, num_chardev):
+        vm.devices.simple_hotplug(char_devices[i], vm.monitor)
+        if i < num_serial_ports:
+            vm.devices.simple_hotplug(serial_devices[i], vm.monitor)
+    for device in serial_devices:
+        add_virtio_ports_to_vm(vm, params, device)
+    if os_type == "windows":
+        driver_name = params["driver_name"]
+        session = vm.wait_for_login()
+        session = utils_test.qemu.windrv_check_running_verifier(
+            session, vm, test, driver_name)
+    run_bg_test()
+    try:
+        out1 = vm.devices.simple_hotplug(serial_devices[0], vm.monitor)
+        out2 = vm.devices.simple_hotplug(char_devices[0], vm.monitor)
+        if out1[1] is True:
+            msg = ("The device %s shoudn't be hotplugged successfully"
+                   % serial_devices[0])
+            test.fail(msg)
+        elif out2[1] is True:
+            msg = ("The device %s shoudn't be hotplugged successfully"
+                   % char_devices[0])
+            test.fail(msg)
+    except DeviceHotplugError:
+        pass
+    if not utils_misc.wait_for(lambda: driver_in_use.check_bg_running(
+            vm, params), check_bg_timeout, 0, 1):
```

Comment:
Do we need to check hot unplug serial and chardev device while transfering data ?
```
6.After data is transferring, 
6.1 hot-plug existed chardevs,virtio serial ports 
6.2 hot-unplug chardevs ,virtio-serial-ports

There should be properly alert message.
```

Reply:
@nanliu-r  You use `join` in the latest code, and the result is same with before:  wait for transfer data finished, then hotunplug device. Is this the expect result ?   From the case step, seems it hotunplug device while transferring data ?

Reply:
> @nanliu-r You use `join` in the latest code, and the result is same with before: wait for transfer data finished, then hotunplug device. Is this the expect result ? From the case step, seems it hotunplug device while transferring data ?

will confirm again.

Reply:
> @nanliu-r You use `join` in the latest code, and the result is same with before: wait for transfer data finished, then hotunplug device. Is this the expect result? From the case step, seems it hotunplug device while transferring data?

Confirmed with xigao, we think that the test point of this case is to ensure every port works well and after transferring data we can still find the port to hot-unplug.
besides, as for the negative test of repeatedly hotplugging port it should test the alert message as you said above, and I think test one port is enough.
since porlarion test case steps are a little confused with this, I will make it more clear in polarion.
Thanks.

Reply:
> @nanliu-r You use `join` in the latest code, and the result is same with before: wait for transfer data finished, then hotunplug device. Is this the expect result ? From the case step, seems it hotunplug device while transferring data ?

In order to test every port smoothly, we should wait for transferring data to finish, and I will update my porlarion.

Reply:
ok, thanks for your confirming.